proposals/p2188.md

```diff
@@ -0,0 +1,655 @@
+  H(1, (2, 3));
+}
+```
+
+This would ensure consistency between pattern matching and deduction,
+potentially reducing the number of rules that Carbon developers need to learn.
+
+We find that attempting to unify pattern-matching and type deduction in this way
+harms readability. Having distinct syntax for type-level matching and
+value-level matching helps guide the reader to the correct interpretation, even
+though the underlying matching process is expected to be similar or identical.
+As a result, we keep type deduction syntax and pattern matching syntax separate
+for now:
+
+-   In pattern matching, bindings and wildcards are introduced by nested `:` /
+    `:!` patterns, and the right-hand side of a binding pattern is never a
+    proper pattern.
+-   In type deduction, deduced values are specified separately and an expression
+    written in terms of those bindings describes the type.
+
+### Introducer syntax for expression patterns
+
+We could have some separate introducer syntax to distinguish expression patterns
+from other kinds of patterns:
+
+```
+match ((a, b)) {
+  case is (1, 2) => {}
+  case (is 3, n: i32) => {}
+  case (m: i32, is 4) => {}
+}
+```
+
+This would reduce the chance of confusion in cases where an expression and a
+similar-looking pattern are treated differently:
+
+```
+class TupleLike {
+  external impl as (i32, i32);
+}
+fn MatchTupleLike(t: TupleLike) {
+  match (t) {
+    // ✅ OK, expression pattern;
+    // `t` implicitly converted to `(i32, i32)` by
+    // built-in `impl EqWith` for tuples.
+    case (1, 2) => {}
+    // ❌ Error, `t` is not a tuple.
+    case (n: i32, 3) => {}
+  }
+}
+```
+
+However, this would also introduce additional ceremony for the common case where
+part of a pattern is a specific value. This could be mitigated by permitting
+certain kinds of value as patterns without an introducer, such as numeric
+literals and `true` and `false`, at the cost of introducing more complexity and
+more confusion over which cases require `is` and which do not.
+
+We have also not identified a good choice for the introducer syntax, should we
+pursue this direction. `is` is not an ideal choice, because elsewhere in Carbon
+syntax, it is a relation between a value and its type, so `is T` may be misread
+as matching values whose type is `T`. `==` _expression_ has been suggested, but
+that would imply the opposite operand order of that in this proposal --
+_scrutinee_ `==` _expression_ rather than _expression_ `==` _scrutinee_ -- which
+would compare struct fields in a surprising order that diverges from the order
+of comparison for a struct pattern.
+
+## Future work
+
+### Or patterns
+
+We could provide "or patterns", allowing matching of one pattern or another with
+the same handler:
+
+```
+match (x) {
+  case (m: i32, 0) | (0, m: i32) => { return m; }
+}
+```
+
+### Guards
+
+We could provide guards for patterns, allowing a pattern to match only if some
+predicate involving the bindings holds:
```

Comment:
A pattern matching syntax without guards is very difficult to use when a guard is needed. It seems to me like this should be part of the base set of operations and its low complexity doesn't seem to warrant an initial exclusion.